### ISSUE TITLE
[bug] conan 1.x vs 2.x differences in new compatibility() method vs package_id() method

### ISSUE THREAD

**Issue by puetzk** [conan-io/conan#12476]
https://docs.conan.io/en/latest/migrating_to_2.0/recipes.html#migrate-conanfile-compatible-packages-to-the-new-compatibility-method recommends migrating recipes from package_id: ... self.compatible_packages.append to the new compatibility() method even in 1.x

However, the behavior of 1.53 and 2.0.0-beta4 do not match, and the 1.x implementation does not seem to be viable if the package has anything non-default going on in its package_id method, e.g. the sort of things shown at https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html?highlight=defining%20abi#defining-a-custom-package-id like deleting options/subsettings, or assigning broader category strings before hashing.

### Environment Details (include every applicable attribute)
  * Operating System+version: Windows 10 20H2
  * Conan version: 1.53.0 / 2.0.0-beta4
  * Python version: 3.10.7

### Steps to reproduce (Include if Applicable)

<details>
<summary>conanfile.py</summary>

```python
from conan import ConanFile

class TestConanFile(ConanFile):
    name = "test"
    version = "0.1"

    options = {
        "compat": ["A", "B", "AB"],
        "info_only": [1,2],
    }

    default_options = {
            "compat": "A",
            "info_only": 1,
    }

    def compatibility(self):
        if self.options.compat == "A" or self.options.compat == "B":
            return [
                    {"options":[
                        ("compat","AB"),
#                      ("info_only",None),
                    ]},
            ]

    def package_id(self):
#        self.output.info("package_id: %s" % self.info.dumps())
        del self.info.options.info_only # option that only affects package_info and doesn't produce a separate package_id
```
</details>

`options.info_only` represents something that would affect package_info but can be disregarded by package_id. E.g. something like [`no_main` in conancenter's gtest package](https://github.com/conan-io/conan-center-index/blob/53701c8d799efbea53aac76528b9cf0cc660d569/recipes/gtest/all/conanfile.py#L85).

### Logs (Executed commands with output) (Include/Attach if Applicable)

`conan export-pkg conanfile.py -o compat=AB`
> test/0.1: Package '2a1d269613fc54bcec85cbcd305de9a83c570bc2' created

`conan info test/0.1@ -o compat=A`
> ID: d11f361b89caf2b5d6f08e4f72c5d2d1b0b4783e

d11f361b89caf2b5d6f08e4f72c5d2d1b0b4783e doesn't exist, and it failed to find the compatible package 2a1d269613fc54bcec85cbcd305de9a83c570bc2 with `options.compat=AB` (despite the compatibility() method). Sticking a `print(compat_info.package_id())` into [_compatible_infos](https://github.com/conan-io/conan/blob/3e08841b55bd7bb83af31d212774b998add503f1/conans/client/graph/compatibility.py#L33) shows that it looked for 51c2ac4602637ed8f01dad0cb6dce1e97f511eea, which would be the hash of a package with compat=AB and info_only=1. But such an id will never exist, because of `package:id(self): del self.info.options.info_only`.

2.0.0-beta4 works:

`conan2 export-pkg conanfile.py -o compat=AB`
`conan2 graph info --require test/0.1 -o compat=A`
> test/0.1: Main binary package 'c9e7469bb04284f8b9e4bb45ea02dd49ddeb848b' missing. Using compatible package '6cbd7cce3e41ee00df89053dfb6e615f18749c3c'

Uncommenting the self.output call in package_id shows that conan 1.53.0 calls package_id(self) for the main binary package, but not for the alternatives it constructed from the tuples returned by compatibility(). Thus, if the package_id deletes or modifies values in `self.info`, no return value from `compatiblility()` would match binary packages that could actually exist.

conan 2.0.0-beta4 calls package_id multiple times (for each info object being considered), computes the correct hashes, and finds the expected match.

Doing things the old way (with self.compatible_packges.append in the package_id method) can work as long as the correct modifications are performed (in this case, simply a matter of doing the `info.clone()` after the `del self.info.options.info_only`)

It does not seem to be possible to make this work even extra with manual care to account for package_id; e.g. uncommenting the `("info_only",None),` tuple errors out showing the values from compatibility are being validated as if they were the initially-bound option values, rather than contents for the `self.info` object.
> ERROR: 'None' is not a valid 'options.info_only' value.
> Possible values are ['1', '2']

2.0 does not seem to perform this validation, and will quietly allow compatibility() to return invalid option values that are passed through to self.info as-is (which package_id can then change further before hashing).

So it seems like neither 1.53.0 nor 2.0.0-beta4 are being consistent here about whether [the return tuple values `compatibility()`](https://docs.conan.io/1/reference/conanfile/methods.html#method-compatibility) are meant to be self.options or self.info.options. 1.53.0 is validating it like --options input from the profile. but then not calling package_id to translate it before hashing. 2.0.0-beta4 is *not* validating the options, but *is* passing through the (possibly-invalid) values when it invokes self.package_id. And the documentation doesn't really say either way which interpretation is intended.

**Comment by puetzk** [comment]
Re-checked and the same behaviors are still present in 1.59.0/2.0.1; in 1.59.0 the options offered by compatibility() are not passed through package_id(), and in 2.0.1 things like info_only=None do not get validated.

**Comment by memsharded** [comment]
Hi @puetzk 

Thanks for your detailed report.
I have been checking this, and indeed it seems there is a gap in the compatibility regarding the removal of options, that will be difficult to bridge. Fortunately it seems the 2.0 behavior is the expected one, isn't it?

I think these could be the options to handle this case:

- Replace the removal in ``package_id()`` method with an assignment to a constant value. The net effect will be the same, always the same ``package_id()`` irrespective of the option value.
  ```python
      def package_id(self):
        self.info.options.info_only = "1"
  ```
- Introduce a conditional in the recipe to the conan version to re-enable and use the older ``compatible_packages`` syntax in ``package_id()`` method.
 ```python
 from conan import ConanFile, conan_version

class TestConanFile(ConanFile):
    name = "test"
    version = "0.1"
    def package_id(self):
        del self.info.options.info_only 
        if conan_version < "2.0":
            # older compatible_packages syntax
 ```

**Comment by puetzk** [comment]
Yes, the 2.0 behavior definitely seems like the right one (in 2.0 the only problem is that it *doesn't* validate options returned by compatibility(), which is harmless as long as the recipe gets its own options spelled right).

I was originally digging at this stuff as offshoot of https://github.com/conan-io/conan/issues/10609 and https://github.com/conan-io/conan/issues/10617 (which also works in 2.x but not 1.x), but I filed it separately after noticing that https://docs.conan.io/1/migrating_to_2.0/recipes.html#migrate-conanfile-compatible-packages-to-the-new-compatibility-method recommends translating `self.compatible_packages.append` to `def compatibility(self)`. Per this bug, that direct translation will often no longer work (deleted options are probably fairly rare, but the same goes for settings to, and things like deleting compiler.cppstd or compiler.libcxx are rather common, especially for the sort of library with tightly-scoped ABI that compatibility is suitable for in the first place).

But I filed the issue more because "this seems unintentional, and likely to bite others" than because I have a specific package where it's a blocker. So maybe all we get is "this works in 2.0 but has known issues in 1.x", and at least now there's an issue documenting that for anyone else who stumbles into it :-)

> Introduce a conditional in the recipe to the conan version to re-enable and use the older compatible_packages syntax in package_id() method.

Maybe, though you still get into the trouble than that 1.x will validate the options returned from compatibility() as if they were user input. So you'd have to make this fixed placeholder one of the "real" choices to avoid that. Kinda gross.


In which case, here's hoping JFrog gets artifactory >= 7.52 (https://github.com/conan-io/conan/issues/11370#issuecomment-1415826616) rolled out on their SaaS artifactory service eventually so I can start looking seriously at our path to migrate to conan 2.0.x beyond some testing. 

Right now I'm still seeing our https://*.customers.jfrog.net/artifactory/api/system/version instance showing as 7.50.7 anduploading any conan 2.0 binaries seemingly breaks search indexing for both conan 1.x and 2.x. I'm assuming this is just RTDEV-27896. but that's got any real movement toward conan 2.0 stymied for now.

**Comment by memsharded** [comment]
> Yes, the 2.0 behavior definitely seems like the right one (in 2.0 the only problem is that it doesn't validate options returned by compatibility(), which is harmless as long as the recipe gets its own options spelled right).

Yes, that was a trade-off we made, and it seemed reasonable to get the core behavior right and not make it more complex or fragile just to try to protect against potential errors.

> Per this bug, that direct translation will often no longer work (deleted options are probably fairly rare, but the same goes for settings to, and things like deleting compiler.cppstd or compiler.libcxx are rather common, especially for the sort of library with tightly-scoped ABI that compatibility is suitable for in the first place).

Deleting ``compiler.cppstd`` and ``compiler.libcxx`` is indeed very common for C libraries, but that shouldn't happen in ``package_id()``, but in ``configure()`` (so it is not applied during ``build()`` either), so this particular case shouldn't be an issue.

But yes, it seems that the case of deletion was overlooked in this case, and now it would be terribly challenging to provide it.

> Maybe, though you still get into the trouble than that 1.x will validate the options returned from compatibility() as if they were user input. So you'd have to make this fixed placeholder one of the "real" choices to avoid that. Kinda gross.

Yes, I wasn't fully detailing the solution. ``compatibility()`` would also be disabled by ``conan_version < 2.0`` and will use the legacy syntax in ``package_id()``. It is quite likely that in any case the best solution in 2.0 could be the global ``compatibility.py`` plugin, and not having to detail things in recipes, which probably would be less frequent.

> In which case, here's hoping JFrog gets artifactory >= 7.52 (https://github.com/conan-io/conan/issues/11370#issuecomment-1415826616) rolled out on their SaaS artifactory service eventually so I can start looking seriously at our path to migrate to conan 2.0.x beyond some testing.

We have been developing Conan 2.0, including preparing ConanCenter using <7.52, and it is possible to do like 98% of the things, because that "search binaries" endpoint is typically not in critical paths (we implemented a quick workaround in one place, but calling Artifactory API to list the subfolders of a path, then getting all package_ids, which are subfolders), all the core important functionality, like uploads, installs, list revisions, etc, was working before. It shouldn't affect search of recipes/versions only listing of binaries.

In any case, I am happy to learn that the several issues, like this one, and https://github.com/conan-io/conan/issues/10609 and https://github.com/conan-io/conan/issues/10617 "(which also works in 2.x but not 1.x)" were correctly addressed in 2.0, and it seems it is in general providing the right solutions for these issues.

**Comment by memsharded** [comment]
I am closing this ticket as outdated/solved in Conan 2, as described above, the issues should be resolved now in Conan 2.
Please create new tickets for any further question or issue. Many thanks for the feedback!
